File: Leetcode/117.populatingNextRightPointersInEachNodeII/117.cpp
 void connect(TreeLinkNode *root) 
{
	if (root == NULL) return;
	TreeLinkNode * head = NULL;// head node of next level
	TreeLinkNode * prev = NULL;//leading node in the next level
	TreeLinkNode * cur = root;// current node of current level 
	while (cur)
	{
		while (cur)
		{
			if (cur -> left)//left child
			{
				if (prev)
					prev -> next = cur -> left;
				else 
					head = cur -> left;// if prev do not exist, we set the head; 
				prev = cur -> left;// prev is is current node's left child.
			}
			if (cur -> right)
			{
				if (prev)
					prev ->next = cur -> right;
				else 
					head = cur -> right;
				prev = cur -> right;
			}
			cur = cur -> next;
		}
		cur = head;
		prev = NULL;
		head = NULL;
	}
}

File: Leetcode/117.populatingNextRightPointersInEachNodeII/117.java
 //https://leetcode.com/discuss/3339/o-1-space-o-n-complexity-iterative-solution
 public class Solution {
        public static void connect(TreeLinkNode root) {
            TreeLinkNode nextHead = new TreeLinkNode(0);
            nextHead.next = root;
            while(nextHead.next != null){
                TreeLinkNode tail = nextHead;
                TreeLinkNode n = nextHead.next;
                nextHead.next = null;
                for(; n != null; n = n.next){
                    if(n.left != null){
                        tail.next = n.left;
                        tail = tail.next;//tail and nexthead are not same now
                    }

                    if(n.right != null){
                        tail.next = n.right; 
                        tail = tail.next;
                    }
                }
            }
        }
    }
